verbs provider fails to compile with gcc 4.4.7 (RHEL 6 default compiler)

This is master head as of 2018-01-11:

```
$ gcc --version
gcc (GCC) 4.4.7 20120313 (Red Hat 4.4.7-4)
Copyright (C) 2010 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

$ make
make  all-am
make[1]: Entering directory `/home/jsquyres/git/libfabric'
  CC       prov/verbs/src/src_libfabric_la-fi_verbs.lo
In file included from prov/verbs/src/fi_verbs.c:37:
prov/verbs/src/fi_verbs.h: In function ‘fi_ibv_mem_notifier_set_free_hook’:
prov/verbs/src/fi_verbs.h:537: error: #pragma GCC diagnostic not allowed inside functions
prov/verbs/src/fi_verbs.h:538: error: #pragma GCC diagnostic not allowed inside functions
prov/verbs/src/fi_verbs.h:540: error: #pragma GCC diagnostic not allowed inside functions
prov/verbs/src/fi_verbs.h: In function ‘fi_ibv_mem_notifier_set_realloc_hook’:
prov/verbs/src/fi_verbs.h:564: error: #pragma GCC diagnostic not allowed inside functions
prov/verbs/src/fi_verbs.h:565: error: #pragma GCC diagnostic not allowed inside functions
prov/verbs/src/fi_verbs.h:567: error: #pragma GCC diagnostic not allowed inside functions
prov/verbs/src/fi_verbs.h: In function ‘fi_ibv_mem_notifier_get_free_hook’:
prov/verbs/src/fi_verbs.h:591: error: #pragma GCC diagnostic not allowed inside functions
prov/verbs/src/fi_verbs.h:592: error: #pragma GCC diagnostic not allowed inside functions
prov/verbs/src/fi_verbs.h:594: error: #pragma GCC diagnostic not allowed inside functions
prov/verbs/src/fi_verbs.h: In function ‘fi_ibv_mem_notifier_get_realloc_hook’:
prov/verbs/src/fi_verbs.h:618: error: #pragma GCC diagnostic not allowed inside functions
prov/verbs/src/fi_verbs.h:619: error: #pragma GCC diagnostic not allowed inside functions
prov/verbs/src/fi_verbs.h:621: error: #pragma GCC diagnostic not allowed inside functions
make[1]: *** [prov/verbs/src/src_libfabric_la-fi_verbs.lo] Error 1
make[1]: Leaving directory `/home/jsquyres/git/libfabric'
make: *** [all] Error 2
```

--- a/prov/verbs/src/fi_verbs.h
+++ b/prov/verbs/src/fi_verbs.h
@@ -533,7 +533,7 @@ fi_ibv_mem_notifier_set_free_hook(fi_ibv_mem_free_hook free_hook)
 #  pragma clang diagnostic ignored "-Wdeprecated-declarations"
 	__free_hook = free_hook;
 #  pragma clang diagnostic pop
-# elif defined __GNUC__ /* GCC */
+# elif __GNUC__ > 4 || (__GNUC__ == 4 && __GNUC_MINOR__ >= 6) /* GCC >= 4.6 */
 #  pragma GCC diagnostic push
 #  pragma GCC diagnostic ignored "-Wdeprecated-declarations"
 	__free_hook = free_hook;
@@ -560,7 +560,7 @@ fi_ibv_mem_notifier_set_realloc_hook(fi_ibv_mem_realloc_hook realloc_hook)
 #  pragma clang diagnostic ignored "-Wdeprecated-declarations"
 	__realloc_hook = realloc_hook;
 #  pragma clang diagnostic pop
-# elif defined __GNUC__ /* GCC */
+# elif __GNUC__ > 4 || (__GNUC__ == 4 && __GNUC_MINOR__ >= 6) /* GCC >= 4.6 */
 #  pragma GCC diagnostic push
 #  pragma GCC diagnostic ignored "-Wdeprecated-declarations"
 	__realloc_hook = realloc_hook;
@@ -587,7 +587,7 @@ fi_ibv_mem_notifier_get_free_hook(void)
 #  pragma clang diagnostic ignored "-Wdeprecated-declarations"
 	return __free_hook;
 #  pragma clang diagnostic pop
-# elif defined __GNUC__ /* GCC */
+# elif __GNUC__ > 4 || (__GNUC__ == 4 && __GNUC_MINOR__ >= 6) /* GCC >= 4.6 */
 #  pragma GCC diagnostic push
 #  pragma GCC diagnostic ignored "-Wdeprecated-declarations"
 	return __free_hook;
@@ -614,7 +614,7 @@ fi_ibv_mem_notifier_get_realloc_hook(void)
 #  pragma clang diagnostic ignored "-Wdeprecated-declarations"
 	return __realloc_hook;
 #  pragma clang diagnostic pop
-# elif defined __GNUC__ /* GCC */
+# elif __GNUC__ > 4 || (__GNUC__ == 4 && __GNUC_MINOR__ >= 6) /* GCC >= 4.6 */
 #  pragma GCC diagnostic push
 #  pragma GCC diagnostic ignored "-Wdeprecated-declarations"
 	return __realloc_hook;

